samcli/lib/telemetry/metric.py

```diff
@@ -29,16 +30,120 @@
 No side effect will result in this as it is write-only for code outside of telemetry.
 Decorators should be used to minimize logic involving telemetry.
 """
-_METRICS = dict()
+_METRICS: Optional[defaultdict] = None
+
+
+class MetricName(Enum):
+    # Metrics that can be reported only once in a session
+    installed = ("installed", True)
+    commandRun = ("commandRun", True)
+
+    # Metrics that can be reported any number of times during a session
+    # e.g. docker_image_download_status
+    templateWarning = ("templateWarning", False)
+    runtimeMetric = ("runtimeMetric", False)
+
+    def __init__(self, name: str, can_report_once_only: bool):
+        self._name = name
+        self.can_report_once_only = can_report_once_only
+
+    def __str__(self) -> str:
+        return self._name
+
+
+def _get_global_metric_sink() -> defaultdict:
+    global _METRICS  # pylint: disable=global-statement
+    if _METRICS is None:
+        _METRICS = defaultdict(list)
+    return _METRICS
+
+
+def flush():
+    """
+    Emits all metrics to telemetry and flush the metric sink
+    """
+    telemetry = Telemetry()
+    metric_sink = _get_global_metric_sink()
+    for metric_list in metric_sink.values():
+        while metric_list:
+            metric = metric_list.pop(0)
```

Comment:
is this stack based implementation because we inherently value the latest information more than what was first tracked within the session? if that is the case, lets reason it out and add a comment, could help shape our design philosphy.

Reply:
There are two cases here:
1. metrics that can be reported at most once per session (e.g. `commandRun`) - in this case, the stack will always be of length one. 
2. metrics that can be reported more than once per session (e.g. `templateWarning` and `runtimeMetric`) - in this case, the `metric_list` works more like a queue than a stack (i.e. FIFO). This will preserve the chronicle order when they are emitted to the telemetry api.

I'll add a comment explaining this explicitly